### Data is about Mice Protein

Wow! it is all about biology where there is all types of test run is performed on a mouse

Coming to the data set:

    It is a multivariate data analysis

    It has 1080 instances and 82 attributes

    The characteristics which is represented in the data set is real and it is performed in the year 2015 August 04
    
    
#### Intially performing EDA on Mice protein 

## Importing the required Libraries

In [1]:
import matplotlib.pyplot as plt
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score 
import pandas as pd


In [2]:
mice= pd.read_csv("mice_protein.csv", sep=",")

In [3]:

pd.options.display.max_colwidth=-1

In [4]:
mice.head()

MouseID     DYRK1A_N      ITSN1_N       BDNF_N        NR1_N       NR2A_N  \
0  '309_1'  0.503643884  0.747193224  0.4301753    2.81632854   5.990151664   
1  '309_2'  0.51461708   0.689063548  0.411770344  2.789514042  5.685037861   
2  '309_3'  0.509183088  0.730246795  0.418308781  2.687201071  5.622058542   
3  '309_4'  0.442106692  0.61707615   0.358626307  2.466947197  4.97950319    
4  '309_5'  0.434940244  0.617429838  0.358802202  2.36578488   4.718678663   

        pAKT_N      pBRAF_N    pCAMKII_N      pCREB_N  ...      pCFOS_N  \
0  0.218830018  0.177565491  2.373744337  0.232223754  ...  0.108335872   
1  0.211636155  0.172817023  2.292149909  0.226972108  ...  0.10431543    
2  0.209010905  0.175722212  2.283336522  0.230246795  ...  0.106219278   
3  0.222885842  0.176462604  2.152300801  0.207004208  ...  0.111261965   
4  0.213105949  0.173626964  2.134013697  0.192157916  ...  0.110693923   

      SYP_N    H3AcK18_N       EGR1_N     H3MeK4_N    CaNA_N   Genotype  \
0  0.427099  0.114783227  0.131790029  0.128185603  1.675652  'Control'   
1  0.441581  0.111973507  0.13510297   0.1311187    1.743610  'Control'   
2  0.435777  0.111882881  0.133361829  0.127431075  1.926427  'Control'   
3  0.391691  0.130405268  0.147444165  0.146901093  1.700563  'Control'   
4  0.434154  0.118481435  0.140314282  0.148379919  1.839730  'Control'   

     Treatment Behavior     class  
0  'Memantine'  'C/S'    'c-CS-m'  
1  'Memantine'  'C/S'    'c-CS-m'  
2  'Memantine'  'C/S'    'c-CS-m'  
3  'Memantine'  'C/S'    'c-CS-m'  
4  'Memantine'  'C/S'    'c-CS-m'  

[5 rows x 82 columns]

### What are the 82 attributes telling us?

1 Mouse ID which can be removed

2 to 78 Values of expression levels of 77 proteins; the names of proteins are followed by “_n” indicating that they were measured in the nuclear fraction. For example: DYRK1A_n\

79th attribute is Genotype: control (c) or trisomy (t) 

80th attribute is Treatment type: memantine (m) or saline (s)

81st attribute is Behavior: context-shock (CS) or shock-context (SC)

82nd attribute is Class: c-CS-s, c-CS-m, c-SC-s, c-SC-m, t-CS-s, t-CS-m, t-SC-s, t-SC-m

so class is my target variable which tells 

### Classes:
c-CS-s: control mice, stimulated to learn, injected with saline 

c-CS-m: control mice, stimulated to learn, injected with memantine 

c-SC-s: control mice, not stimulated to learn, injected with saline 

c-SC-m: control mice, not stimulated to learn, injected with memantine 

t-CS-s: trisomy mice, stimulated to learn, injected with saline 

t-CS-m: trisomy mice, stimulated to learn, injected with memantine 

t-SC-s: trisomy mice, not stimulated to learn, injected with saline

t-SC-m: trisomy mice, not stimulated to learn, injected with memantine 



In [5]:
mice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1080 entries, 0 to 1079
Data columns (total 82 columns):
MouseID            1080 non-null object
DYRK1A_N           1080 non-null object
ITSN1_N            1080 non-null object
BDNF_N             1080 non-null object
NR1_N              1080 non-null object
NR2A_N             1080 non-null object
pAKT_N             1080 non-null object
pBRAF_N            1080 non-null object
pCAMKII_N          1080 non-null object
pCREB_N            1080 non-null object
pELK_N             1080 non-null object
pERK_N             1080 non-null object
pJNK_N             1080 non-null object
PKCA_N             1080 non-null object
pMEK_N             1080 non-null object
pNR1_N             1080 non-null object
pNR2A_N            1080 non-null object
pNR2B_N            1080 non-null object
pPKCAB_N           1080 non-null object
pRSK_N             1080 non-null object
AKT_N              1080 non-null object
BRAF_N             1080 non-null object
CAMKII_N     

As we see the data set tells there is no missing values but according to the data where i got form "https://www.openml.org/d/40966" tells that there are missing values 

So, let's eyball the datasets

In [6]:
mice.isnull().sum()

MouseID            0
DYRK1A_N           0
ITSN1_N            0
BDNF_N             0
NR1_N              0
NR2A_N             0
pAKT_N             0
pBRAF_N            0
pCAMKII_N          0
pCREB_N            0
pELK_N             0
pERK_N             0
pJNK_N             0
PKCA_N             0
pMEK_N             0
pNR1_N             0
pNR2A_N            0
pNR2B_N            0
pPKCAB_N           0
pRSK_N             0
AKT_N              0
BRAF_N             0
CAMKII_N           0
CREB_N             0
ELK_N              0
ERK_N              0
GSK3B_N            0
JNK_N              0
MEK_N              0
TRKA_N             0
                  ..
RRP1_N             0
BAX_N              0
ARC_N              0
ERBB4_N            0
nNOS_N             0
Tau_N              0
GFAP_N             0
GluR3_N            0
GluR4_N            0
IL1B_N             0
P3525_N            0
pCASP9_N           0
PSD95_N            0
SNCA_N             0
Ubiquitin_N        0
pGSK3B_Tyr216_N    0
SHH_N        

Yes according to the jupyter notebook we don't have any missing values as the value is "?" it is considered as a category

In [7]:
missing_values= ['?']
mice= pd.read_csv("mice_protein.csv", sep=",", na_values = missing_values)

In [8]:
mice.head()

MouseID  DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N  \
0  '309_1'  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830   
1  '309_2'  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636   
2  '309_3'  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011   
3  '309_4'  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886   
4  '309_5'  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106   

    pBRAF_N  pCAMKII_N   pCREB_N  ...   pCFOS_N     SYP_N  H3AcK18_N  \
0  0.177565  2.373744   0.232224  ...  0.108336  0.427099  0.114783    
1  0.172817  2.292150   0.226972  ...  0.104315  0.441581  0.111974    
2  0.175722  2.283337   0.230247  ...  0.106219  0.435777  0.111883    
3  0.176463  2.152301   0.207004  ...  0.111262  0.391691  0.130405    
4  0.173627  2.134014   0.192158  ...  0.110694  0.434154  0.118481    

     EGR1_N  H3MeK4_N    CaNA_N   Genotype    Treatment  Behavior     class  
0  0.131790  0.128186  1.675652  'Control'  'Memantine'  'C/S'     'c-CS-m'  
1  0.135103  0.131119  1.743610  'Control'  'Memantine'  'C/S'     'c-CS-m'  
2  0.133362  0.127431  1.926427  'Control'  'Memantine'  'C/S'     'c-CS-m'  
3  0.147444  0.146901  1.700563  'Control'  'Memantine'  'C/S'     'c-CS-m'  
4  0.140314  0.148380  1.839730  'Control'  'Memantine'  'C/S'     'c-CS-m'  

[5 rows x 82 columns]

Now we have missing values in the name of "NaN"

In [9]:
mice.isnull().sum()

MouseID            0  
DYRK1A_N           3  
ITSN1_N            3  
BDNF_N             3  
NR1_N              3  
NR2A_N             3  
pAKT_N             3  
pBRAF_N            3  
pCAMKII_N          3  
pCREB_N            3  
pELK_N             3  
pERK_N             3  
pJNK_N             3  
PKCA_N             3  
pMEK_N             3  
pNR1_N             3  
pNR2A_N            3  
pNR2B_N            3  
pPKCAB_N           3  
pRSK_N             3  
AKT_N              3  
BRAF_N             3  
CAMKII_N           3  
CREB_N             3  
ELK_N              18 
ERK_N              3  
GSK3B_N            3  
JNK_N              3  
MEK_N              7  
TRKA_N             3  
                  ..  
RRP1_N             0  
BAX_N              0  
ARC_N              0  
ERBB4_N            0  
nNOS_N             0  
Tau_N              0  
GFAP_N             0  
GluR3_N            0  
GluR4_N            0  
IL1B_N             0  
P3525_N            0  
pCASP9_N           0  
PSD95_N    

now we have missing values

### Step1: Fixing the Missing values

In [10]:
mice.columns=mice.columns.str.lower()

In [11]:
mice.columns

Index(['mouseid', 'dyrk1a_n', 'itsn1_n', 'bdnf_n', 'nr1_n', 'nr2a_n', 'pakt_n',
       'pbraf_n', 'pcamkii_n', 'pcreb_n', 'pelk_n', 'perk_n', 'pjnk_n',
       'pkca_n', 'pmek_n', 'pnr1_n', 'pnr2a_n', 'pnr2b_n', 'ppkcab_n',
       'prsk_n', 'akt_n', 'braf_n', 'camkii_n', 'creb_n', 'elk_n', 'erk_n',
       'gsk3b_n', 'jnk_n', 'mek_n', 'trka_n', 'rsk_n', 'app_n', 'bcatenin_n',
       'sod1_n', 'mtor_n', 'p38_n', 'pmtor_n', 'dscr1_n', 'ampka_n', 'nr2b_n',
       'pnumb_n', 'raptor_n', 'tiam1_n', 'pp70s6_n', 'numb_n', 'p70s6_n',
       'pgsk3b_n', 'ppkcg_n', 'cdk5_n', 's6_n', 'adarb1_n', 'acetylh3k9_n',
       'rrp1_n', 'bax_n', 'arc_n', 'erbb4_n', 'nnos_n', 'tau_n', 'gfap_n',
       'glur3_n', 'glur4_n', 'il1b_n', 'p3525_n', 'pcasp9_n', 'psd95_n',
       'snca_n', 'ubiquitin_n', 'pgsk3b_tyr216_n', 'shh_n', 'bad_n', 'bcl2_n',
       'ps6_n', 'pcfos_n', 'syp_n', 'h3ack18_n', 'egr1_n', 'h3mek4_n',
       'cana_n', 'genotype', 'treatment', 'behavior', 'class'],
      dtype='object')

In [12]:
type(np.NaN)

float

As all the missing values are numeric I can use mean/median/mode. mostly it looks data which is familar as a chance I am imparting mean to cover the inequality in the dataset

In [13]:
mice['dyrk1a_n'].describe()

count    1077.000000
mean     0.425810   
std      0.249362   
min      0.145327   
25%      0.288121   
50%      0.366378   
75%      0.487711   
max      2.516367   
Name: dyrk1a_n, dtype: float64

In [14]:
mice['dyrk1a_n'].fillna(mice['dyrk1a_n'].mean(), inplace=True)

In [15]:
mi=mice

I am just trying as  i have 45 columns as missing values i cannot impart or fix the missing value one by one.  So, let me do it at once

In [16]:
mi.fillna(mi.mean(),inplace= True)

In [17]:
mi.isnull().sum()

mouseid            0
dyrk1a_n           0
itsn1_n            0
bdnf_n             0
nr1_n              0
nr2a_n             0
pakt_n             0
pbraf_n            0
pcamkii_n          0
pcreb_n            0
pelk_n             0
perk_n             0
pjnk_n             0
pkca_n             0
pmek_n             0
pnr1_n             0
pnr2a_n            0
pnr2b_n            0
ppkcab_n           0
prsk_n             0
akt_n              0
braf_n             0
camkii_n           0
creb_n             0
elk_n              0
erk_n              0
gsk3b_n            0
jnk_n              0
mek_n              0
trka_n             0
                  ..
rrp1_n             0
bax_n              0
arc_n              0
erbb4_n            0
nnos_n             0
tau_n              0
gfap_n             0
glur3_n            0
glur4_n            0
il1b_n             0
p3525_n            0
pcasp9_n           0
psd95_n            0
snca_n             0
ubiquitin_n        0
pgsk3b_tyr216_n    0
shh_n        

In [18]:
mi.head()

mouseid  dyrk1a_n   itsn1_n    bdnf_n     nr1_n    nr2a_n    pakt_n  \
0  '309_1'  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830   
1  '309_2'  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636   
2  '309_3'  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011   
3  '309_4'  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886   
4  '309_5'  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106   

    pbraf_n  pcamkii_n   pcreb_n  ...   pcfos_n     syp_n  h3ack18_n  \
0  0.177565  2.373744   0.232224  ...  0.108336  0.427099  0.114783    
1  0.172817  2.292150   0.226972  ...  0.104315  0.441581  0.111974    
2  0.175722  2.283337   0.230247  ...  0.106219  0.435777  0.111883    
3  0.176463  2.152301   0.207004  ...  0.111262  0.391691  0.130405    
4  0.173627  2.134014   0.192158  ...  0.110694  0.434154  0.118481    

     egr1_n  h3mek4_n    cana_n   genotype    treatment  behavior     class  
0  0.131790  0.128186  1.675652  'Control'  'Memantine'  'C/S'     'c-CS-m'  
1  0.135103  0.131119  1.743610  'Control'  'Memantine'  'C/S'     'c-CS-m'  
2  0.133362  0.127431  1.926427  'Control'  'Memantine'  'C/S'     'c-CS-m'  
3  0.147444  0.146901  1.700563  'Control'  'Memantine'  'C/S'     'c-CS-m'  
4  0.140314  0.148380  1.839730  'Control'  'Memantine'  'C/S'     'c-CS-m'  

[5 rows x 82 columns]

let me drop mouse id it is not going to help us atall

In [19]:
mi=mi.drop(columns="mouseid")

In [20]:
mi.head()

dyrk1a_n   itsn1_n    bdnf_n     nr1_n    nr2a_n    pakt_n   pbraf_n  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pcamkii_n   pcreb_n    pelk_n  ...   pcfos_n     syp_n  h3ack18_n  \
0  2.373744   0.232224  1.750936  ...  0.108336  0.427099  0.114783    
1  2.292150   0.226972  1.596377  ...  0.104315  0.441581  0.111974    
2  2.283337   0.230247  1.561316  ...  0.106219  0.435777  0.111883    
3  2.152301   0.207004  1.595086  ...  0.111262  0.391691  0.130405    
4  2.134014   0.192158  1.504230  ...  0.110694  0.434154  0.118481    

     egr1_n  h3mek4_n    cana_n   genotype    treatment  behavior     class  
0  0.131790  0.128186  1.675652  'Control'  'Memantine'  'C/S'     'c-CS-m'  
1  0.135103  0.131119  1.743610  'Control'  'Memantine'  'C/S'     'c-CS-m'  
2  0.133362  0.127431  1.926427  'Control'  'Memantine'  'C/S'     'c-CS-m'  
3  0.147444  0.146901  1.700563  'Control'  'Memantine'  'C/S'     'c-CS-m'  
4  0.140314  0.148380  1.839730  'Control'  'Memantine'  'C/S'     'c-CS-m'  

[5 rows x 81 columns]

In [21]:
mi.var()

dyrk1a_n           0.062009
itsn1_n            0.063147
bdnf_n             0.002432
nr1_n              0.120277
nr2a_n             0.868255
pakt_n             0.001729
pbraf_n            0.000729
pcamkii_n          1.672800
pcreb_n            0.001059
pelk_n             0.217393
perk_n             0.118907
pjnk_n             0.002694
pkca_n             0.002721
pmek_n             0.002125
pnr1_n             0.013878
pnr2a_n            0.035251
pnr2b_n            0.073095
ppkcab_n           0.231467
prsk_n             0.004433
akt_n              0.016194
braf_n             0.046694
camkii_n           0.002733
creb_n             0.000693
elk_n              0.110871
erk_n              0.425741
gsk3b_n            0.059737
jnk_n              0.001146
mek_n              0.001676
trka_n             0.014558
rsk_n              0.000790
                     ...   
cdk5_n             0.001397
s6_n               0.018890
adarb1_n           0.130858
acetylh3k9_n       0.034339
rrp1_n             0

In [22]:
mi["genotype"].value_counts()

'Control'    570
'Ts65Dn'     510
Name: genotype, dtype: int64

In [23]:
mi['treatment'].value_counts()

'Memantine'    570
'Saline'       510
Name: treatment, dtype: int64

In [24]:
mi["behavior"].value_counts()

'S/C'    555
'C/S'    525
Name: behavior, dtype: int64

In [25]:
mi["class"].value_counts()

'c-SC-m'    150
'c-CS-m'    150
't-CS-m'    135
't-SC-s'    135
't-SC-m'    135
'c-SC-s'    135
'c-CS-s'    135
't-CS-s'    105
Name: class, dtype: int64

#### So everything has 2-2 classification for every categorical value but only for class there are 8 varibales

It seems that all the genotype, treatment and behavior effects the class variable and the combination of these 3 is the class variable so we can drop all these columns

In [26]:
mi=mi.drop(columns=['genotype','treatment','behavior'])

In [27]:
mi.head()

dyrk1a_n   itsn1_n    bdnf_n     nr1_n    nr2a_n    pakt_n   pbraf_n  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pcamkii_n   pcreb_n    pelk_n  ...     bad_n    bcl2_n     ps6_n   pcfos_n  \
0  2.373744   0.232224  1.750936  ...  0.122652  0.134762  0.106305  0.108336   
1  2.292150   0.226972  1.596377  ...  0.116682  0.134762  0.106592  0.104315   
2  2.283337   0.230247  1.561316  ...  0.118508  0.134762  0.108303  0.106219   
3  2.152301   0.207004  1.595086  ...  0.132781  0.134762  0.103184  0.111262   
4  2.134014   0.192158  1.504230  ...  0.129954  0.134762  0.104784  0.110694   

      syp_n  h3ack18_n    egr1_n  h3mek4_n    cana_n     class  
0  0.427099  0.114783   0.131790  0.128186  1.675652  'c-CS-m'  
1  0.441581  0.111974   0.135103  0.131119  1.743610  'c-CS-m'  
2  0.435777  0.111883   0.133362  0.127431  1.926427  'c-CS-m'  
3  0.391691  0.130405   0.147444  0.146901  1.700563  'c-CS-m'  
4  0.434154  0.118481   0.140314  0.148380  1.839730  'c-CS-m'  

[5 rows x 78 columns]

#### converting the variable names to the range 1-8

In [28]:
target = {"'c-CS-m'": 1, "'c-SC-m'": 2,
      "'c-SC-s'": 3, "'t-CS-m'": 4,
      "'t-SC-m'": 5, "'t-SC-s'": 6,
      "'c-CS-s'": 7, "'t-CS-s'": 8}
mi['class']= [target[i] for i in mi['class']] 
print(mi) 

      dyrk1a_n   itsn1_n    bdnf_n     nr1_n    nr2a_n    pakt_n   pbraf_n  \
0     0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1     0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2     0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3     0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4     0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   
5     0.447506  0.628176  0.367388  2.385939  4.807635  0.218578  0.176233   
6     0.428033  0.573696  0.342709  2.334224  4.473130  0.225173  0.184004   
7     0.416923  0.564036  0.327703  2.260135  4.268735  0.214834  0.179668   
8     0.386311  0.538428  0.317720  2.125725  4.063950  0.207222  0.167778   
9     0.380827  0.499294  0.362462  2.096266  3.598587  0.227649  0.188093   
10    0.366511  0.513278  0.327792  2.072572  3.661058  0.229475  0.193797   
11    0.364154  0.499411  0.355124  2.006871  3.466627  0.216333

In [29]:
mi.head()

dyrk1a_n   itsn1_n    bdnf_n     nr1_n    nr2a_n    pakt_n   pbraf_n  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pcamkii_n   pcreb_n    pelk_n  ...     bad_n    bcl2_n     ps6_n   pcfos_n  \
0  2.373744   0.232224  1.750936  ...  0.122652  0.134762  0.106305  0.108336   
1  2.292150   0.226972  1.596377  ...  0.116682  0.134762  0.106592  0.104315   
2  2.283337   0.230247  1.561316  ...  0.118508  0.134762  0.108303  0.106219   
3  2.152301   0.207004  1.595086  ...  0.132781  0.134762  0.103184  0.111262   
4  2.134014   0.192158  1.504230  ...  0.129954  0.134762  0.104784  0.110694   

      syp_n  h3ack18_n    egr1_n  h3mek4_n    cana_n  class  
0  0.427099  0.114783   0.131790  0.128186  1.675652  1      
1  0.441581  0.111974   0.135103  0.131119  1.743610  1      
2  0.435777  0.111883   0.133362  0.127431  1.926427  1      
3  0.391691  0.130405   0.147444  0.146901  1.700563  1      
4  0.434154  0.118481   0.140314  0.148380  1.839730  1      

[5 rows x 78 columns]

In [30]:
mi['class'].value_counts()

2    150
1    150
7    135
6    135
5    135
4    135
3    135
8    105
Name: class, dtype: int64

 ## OneHotEncodeing 
Not every variable or attribute in this is a numerical variable and before performing noramalization, we have convert the variables into OneHotEncoder 

In [31]:
from sklearn.preprocessing import OneHotEncoder


In [32]:
one = OneHotEncoder(categorical_features = [8,8])


In [33]:
mipro = one.fit_transform(mi).toarray()


C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:392: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [34]:
mipro= pd.DataFrame(mipro)

In [35]:
mipro.head()

0         1         2         3         4         5         6         7  \
0  1.0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  1.0  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  1.0  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  1.0  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  1.0  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

          8         9  ...        68        69        70        71        72  \
0  2.373744  1.750936  ...  0.122652  0.134762  0.106305  0.108336  0.427099   
1  2.292150  1.596377  ...  0.116682  0.134762  0.106592  0.104315  0.441581   
2  2.283337  1.561316  ...  0.118508  0.134762  0.108303  0.106219  0.435777   
3  2.152301  1.595086  ...  0.132781  0.134762  0.103184  0.111262  0.391691   
4  2.134014  1.504230  ...  0.129954  0.134762  0.104784  0.110694  0.434154   

         73        74        75        76   77  
0  0.114783  0.131790  0.128186  1.675652  1.0  
1  0.111974  0.135103  0.131119  1.743610  1.0  
2  0.111883  0.133362  0.127431  1.926427  1.0  
3  0.130405  0.147444  0.146901  1.700563  1.0  
4  0.118481  0.140314  0.148380  1.839730  1.0  

[5 rows x 78 columns]

In [36]:
mipro[77].value_counts()

2.0    150
1.0    150
6.0    135
5.0    135
4.0    135
3.0    135
7.0    135
8.0    105
Name: 77, dtype: int64

## Performing noramalisation using minmaxscaler 

In [37]:
from sklearn.preprocessing import MinMaxScaler

In [38]:
MinMaxScaler(feature_range=(0, 1), copy=True)

MinMaxScaler(copy=True, feature_range=(0, 1))

In [39]:
minmax=MinMaxScaler(feature_range=(0, 1), copy=True)

In [40]:
x = mipro.iloc[:,:-1].values
y = mipro.iloc[:,-1].values

In [41]:
x=minmax.fit_transform(x)
x

array([[0.        , 0.1511224 , 0.21288505, ..., 0.10288975, 0.08457952,
        0.70573764],
       [0.        , 0.1557504 , 0.18822566, ..., 0.11587379, 0.09397699,
        0.74977105],
       [0.        , 0.15345859, 0.20569615, ..., 0.10904993, 0.08216206,
        0.86822857],
       ...,
       [0.        , 0.03516306, 0.06355603, ..., 0.48463073, 0.81196194,
        0.54710006],
       [0.        , 0.03201796, 0.07107097, ..., 0.57133648, 0.84445032,
        0.52973884],
       [0.        , 0.06634184, 0.09150314, ..., 0.57791409, 0.84420951,
        0.50833567]])

In [42]:
np.asarray(x, dtype=float)

array([[0.        , 0.1511224 , 0.21288505, ..., 0.10288975, 0.08457952,
        0.70573764],
       [0.        , 0.1557504 , 0.18822566, ..., 0.11587379, 0.09397699,
        0.74977105],
       [0.        , 0.15345859, 0.20569615, ..., 0.10904993, 0.08216206,
        0.86822857],
       ...,
       [0.        , 0.03516306, 0.06355603, ..., 0.48463073, 0.81196194,
        0.54710006],
       [0.        , 0.03201796, 0.07107097, ..., 0.57133648, 0.84445032,
        0.52973884],
       [0.        , 0.06634184, 0.09150314, ..., 0.57791409, 0.84420951,
        0.50833567]])

In [43]:
type(x)

numpy.ndarray

In [44]:
y

array([1., 1., 1., ..., 6., 6., 6.])

In [45]:
y=np.reshape(y,newshape= (-1,1))
y

array([[1.],
       [1.],
       [1.],
       ...,
       [6.],
       [6.],
       [6.]])

In [46]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(x,y,test_size=0.20,random_state=0)

## Classifing the variables using Logistic Regression

In [47]:
classifier=LogisticRegression(random_state=0)
classifier.fit(X_train,Y_train)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=0, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [48]:
y_pred=classifier.predict(X_test)

In [49]:
accuracy = accuracy_score(Y_test,y_pred)

print(accuracy)

0.8842592592592593


Accuracy has led to 88% which tells that the model is good enough but I need 90% or above so let me do this using Neural networks

## Using of Neural Networks 

In [50]:
from sklearn.neural_network import MLPClassifier

model = MLPClassifier(hidden_layer_sizes=(200,100), max_iter=1000)
model.fit(x, y)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(200, 100), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [51]:
predicted_class=model.predict(X_test)

acneu = accuracy_score(Y_test,predicted_class)

print(acneu)

1.0


Wow just by using neural networks with the hidden layer size of 200, 100 with max_iterations=1000. I got the accuracy of 100%

It seems to be different let me try with different models of hidden layers 

In [52]:
model1 = MLPClassifier(hidden_layer_sizes=(100,200), max_iter=1000)
model1.fit(x, y)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100, 200), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [53]:
predicted_class1=model1.predict(X_test)

acneu1 = accuracy_score(Y_test,predicted_class1)

print(acneu1)

1.0


In [54]:
model2 = MLPClassifier(hidden_layer_sizes=(30,60), max_iter=1000)
model2.fit(x, y)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(30, 60), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [55]:
predicted_class2=model1.predict(X_test)

acneu2 = accuracy_score(Y_test,predicted_class2)
print(acneu2)

1.0


wierd let it be just 30,60 hidden layers even then the accuracy is 100%

## let me change the "test size" 
This may be I can get less than 100% because nothing is perfect and there is no 100 accuracy there will be 0.0003% chance of deviation even if you follow SIX SIGMA level 

In [56]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=0)

In [57]:
classify=LogisticRegression(random_state=0)
classify.fit(x_train,y_train)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=0, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [58]:
Y_pred=classify.predict(x_test)

In [59]:
accuracy1 = accuracy_score(y_test,Y_pred)

print(accuracy1)

0.8740740740740741


ok now my accuarcy has come down from 0.88 to 0.87. so may be i can get a good accuracy even with neural networks

In [60]:
model = MLPClassifier(hidden_layer_sizes=(200,100), max_iter=1000)
model.fit(x, y)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(200, 100), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [61]:
predict_class=model.predict(x_test)

accneu = accuracy_score(y_test,predict_class)

print(accneu)

1.0


In [62]:
model2 = MLPClassifier(hidden_layer_sizes=(30,60), max_iter=1000)
model2.fit(x, y)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(30, 60), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [63]:
predict_class2=model2.predict(x_test)

accneu2 = accuracy_score(y_test,predict_class2)

print(accneu2)

1.0


Again same how is it possible

## using SVM in various ways

In [64]:
from sklearn.svm import SVC
svclassifier = SVC(kernel='poly', degree= 3)# polynomial 
svclassifier.fit(X_train, Y_train)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\svm\base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='poly', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [65]:
y_pred = svclassifier.predict(X_test)


In [66]:
acc = accuracy_score(Y_test,y_pred)
print(acc)

0.1111111111111111


In [67]:
from sklearn.metrics import classification_report
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

         1.0       0.11      1.00      0.20        24
         2.0       0.00      0.00      0.00        25
         3.0       0.00      0.00      0.00        31
         4.0       0.00      0.00      0.00        32
         5.0       0.00      0.00      0.00        24
         6.0       0.00      0.00      0.00        26
         7.0       0.00      0.00      0.00        27
         8.0       0.00      0.00      0.00        27

   micro avg       0.11      0.11      0.11       216
   macro avg       0.01      0.12      0.02       216
weighted avg       0.01      0.11      0.02       216



C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [68]:
svclassifier = SVC(kernel='sigmoid')#sigmoid function
svclassifier.fit(X_train, Y_train)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\svm\base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='sigmoid', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [69]:
y_predsig = svclassifier.predict(X_test)

accsig = accuracy_score(Y_test,y_predsig)
print(accsig)


0.26851851851851855


In [70]:
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

         1.0       0.11      1.00      0.20        24
         2.0       0.00      0.00      0.00        25
         3.0       0.00      0.00      0.00        31
         4.0       0.00      0.00      0.00        32
         5.0       0.00      0.00      0.00        24
         6.0       0.00      0.00      0.00        26
         7.0       0.00      0.00      0.00        27
         8.0       0.00      0.00      0.00        27

   micro avg       0.11      0.11      0.11       216
   macro avg       0.01      0.12      0.02       216
weighted avg       0.01      0.11      0.02       216



C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [71]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.15,random_state=0)

In [72]:
classify=LogisticRegression(random_state=0)
classify.fit(x_train,y_train)

C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Nivya Sree\Anaconda3\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=0, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [73]:
Y_pred=classify.predict(x_test)

accuracy2 = accuracy_score(y_test,Y_pred)

print(accuracy2)

0.8950617283950617


so nice still even when i reduced the text size to 15 then i got the accuracy to 89.5 which i can see to 90% but in the trail and test if it is 89.5% then it is 89.5% cannot be taken as 90%